elfpy/markets/hyperdrive/hyperdrive_actions.py

```diff
@@ -385,28 +578,191 @@ def calc_close_long(
+            # the withdrawal was initiated.
+            lp_margin / open_share_price,
+            lp_interest,
+        )
+        withdrawal_proceeds = withdraw_pool_deltas.withdraw_capital + withdraw_pool_deltas.withdraw_interest
+        share_adjustment -= withdrawal_proceeds
+
+    # Remove the flat component of the trade as well as any LP proceeds paid to the withdrawal pool
+    # from the pool's liquidity.
+    share_reserves = market.market_state.share_reserves + share_reserves_delta
+    bond_reserves = market.market_state.bond_reserves + bond_reserves_delta
+    adjusted_share_reserves, adjusted_bond_reserves = calc_update_reserves(
+        share_reserves, bond_reserves, share_adjustment
+    )
+    share_reserves_delta = adjusted_share_reserves - market.market_state.share_reserves
+    bond_reserves_delta = adjusted_bond_reserves - market.market_state.bond_reserves
+
     # Return the market and wallet deltas.
     market_deltas = MarketDeltas(
-        d_base_asset=trade_result.market_result.d_base,
-        d_bond_asset=trade_result.market_result.d_bonds,
+        d_base_asset=share_reserves_delta * market.market_state.share_price,
+        d_bond_asset=bond_reserves_delta,
         d_base_buffer=-bond_amount,
         long_base_volume=d_base_volume,
         longs_outstanding=-bond_amount,
         long_average_maturity_time=d_long_average_maturity_time,
         long_checkpoints=d_checkpoints,
         total_supply_longs=defaultdict(float, {mint_time: -bond_amount}),
+        withdraw_capital=withdraw_pool_deltas.withdraw_capital,
+        withdraw_interest=withdraw_pool_deltas.withdraw_interest,
+        withdraw_shares_ready_to_withdraw=withdraw_pool_deltas.withdraw_shares_ready_to_withdraw,
     )
     agent_deltas = wallet.Wallet(
         address=wallet_address,
-        balance=types.Quantity(amount=trade_result.user_result.d_base, unit=types.TokenType.BASE),
-        longs={mint_time: wallet.Long(trade_result.user_result.d_bonds)},
-        fees_paid=trade_result.breakdown.fee,
+        balance=types.Quantity(amount=base_proceeds, unit=types.TokenType.BASE),
+        longs={mint_time: wallet.Long(-bond_amount)},
+        fees_paid=fee,
     )
     return market_deltas, agent_deltas
 
 
+def calc_update_reserves(share_reserves: float, bond_reserves: float, share_reserves_delta) -> tuple[float, float]:
+    """Calculates updates to the pool's liquidity and holds the pool's APR constant.
+
+    Parameters
+    ----------
+    share_reserves: float
+        The current total shares in reserve
+    bond_reserves: float
+        The current total bonds in reserve
+    share_reserves_delta:
+        The delta that should be applied to share reserves.
+    """
+    if share_reserves_delta == 0:
+        return 0, 0
+
+    updated_share_reserves = share_reserves + share_reserves_delta
+    updated_bond_reserves = bond_reserves * updated_share_reserves / share_reserves
```

Comment:
I think this is equal to: `updated_bond_reserves = bond_reserves * (1 + share_reserves_delta)`

But, the bond reserves are in units of base, right? So wouldn't you need to convert share reserves delta to base?

Reply:
almost, it's equal to `updated_bond_reserves = bond_reserves * (1 + share_reserves_delta/share_reserves)`, which is just a unitless scalar.